Звездин Роман

Source34 opened the pull request.

Source34 committed:
Initial commit

1) Create solution stucture
2) Add TDD testing for CircularCloudLayouter

Source34 committed:
ArchimedeanSpiral. Declare and TDD

1) Declared ArchimedeanSpiral methods
2) Add TDD tests for ArchimedeanSpiral

Source34 committed:
Busines logic implementation

1) Extracted 3 interfaces ICircularCloud, IImgSaver, IArchimedeanSpiral
2) Add extensions for Rectangle
3) Implemented Circularlayout and ArchimedeanSpiral
4) Implemented BitmapSaver
5) Implemented CircularCloudDrawScenarios, DrawScenario

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,73 @@
+        [TestCase(5, -5, TestName = "Constructor. x > 0 and y < 0")]
+        public void Constructor_WhenIncorrectArgs_ShouldThrow(int x, int y)
+        {
+            Assert.Throws<ArgumentException>(() => new CircularCloudLayouter(new Point(x, y)));
```
Можно переписать на FluentAssertions через `Should().Throws()`

Danko19 replied:
Предлагаю использовать **FluentAssertions** во всех проверках, для единообразия

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core/Helpers/DrawScenario.cs:
```diff
@@ -0,0 +1,63 @@
+        public string ImageName { get; set; }
+        public string SavePath { get; set; }
+
+        public DrawScenario(int rectanglesCount, int maxHeight, int minHeight, int maxWidth, int minWidth, string imageName, Size imageSize)
```
Конструктор кажется лишним, нигде не вызывается. А вообще его можно было через вызов вызов другого

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core/Interfaces/IArchimedeanSpiral.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using System.Drawing;
+
+namespace TagsCloudVisualization.Core.Interfaces
```
Интерфейсы (этот и остальные) кажутся лишними. Мы их объявляем, реализуем, но в местах использования везде ссылаемся на конкретные реализации

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,73 @@
+
+namespace TagsCloudVisualization.Core.Tests
+{
+    public class CircularCloudLayouterTests
```
Не для всех требований из задания есть тесты

Danko19 replied:
Насколько я понял, на каждое требование нужно попробовать написать тест
![image](https://user-images.githubusercontent.com/37505721/201466266-a5418cd0-7803-4c96-a3ad-13332e031641.png)

Source34 replied:
Добавил недостающие тесты

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,73 @@
+            {
+                foreach (var rect2 in circularCloud.Rectangles)
+                {
+                    Assert.False(rect.IntersectsWith(rect2));
```
Этот тест не проходит, падает

Source34 replied:
++

Source34 committed:
Fixes

1) Add README.md with sapmles
2) Add TearDown in tests
3) Remove redundant intefaces and code

Source34 committed:
Update README.md

Source34 committed:
Update CircularCloudLayouterTests

1) Add tests and cases

Source34 committed:
Update CircularCloudLayouterTests

1) Fix img saving when error case in tests

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,108 @@
+{
+    public class CircularCloudLayouterTests
+    {
+        private Point _center;
```
`_center` используется только при создании `CircularCloudLayouter`, кажется можно обойтись локальной переменной

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,108 @@
+        private CircularCloudLayouter _circularCloudLayouter;
+
+        [SetUp]
+        public void Setup()
```
Такой SetUp кажется не очень нужным, потому что в половине все равно тестов создаётся свой `CircularCloudLayouter`. Предлагаю создавать его в тестах всегда

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,108 @@
+        }
+
+        [TestCase(1200, 1200, 50, TestName = "PutNextRectangle. Check rect intersection")]
+        public void PutNextRectangle_RectstIsNotIntersects_ShouldBeTrue(int height, int width, int rectCount)
```
Небольшая опечаточка в названии 😏

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,108 @@
+            var rnd = new Random();
+            var circularCloud = new CircularCloudLayouter(new Point(height / 2, width / 2));
+
+            var rectangles = new List<Rectangle>();
```
Точно нужно отдельно копить `rectngles`? У нас ведь есть `circularCloud.Rectangles`

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,108 @@
+                rectangles.Add(circularCloud.PutNextRectangle(size));
+            }
+
+            foreach (var rect in rectangles)
```
Код через 2 foreach-а казался понятнее 😅

Source34 replied:
++ Предложил немного другое решение, но уверен читаемость возросла :)

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,108 @@
+        public void PutNextRectangle_RectstIsNotIntersects_ShouldBeTrue(int height, int width, int rectCount)
+        {
+            var rnd = new Random();
+            var circularCloud = new CircularCloudLayouter(new Point(height / 2, width / 2));
```
Из-за того, что создаём локальный `circularCloud` в случае падения картинка не рисуется :(

Source34 replied:
++

Source34 committed:
Update CircularCloudLayouterTests

1) Refactor tests
2) Fix cases namig

Source34 committed:
Upadate CircularCloudLayoter

1) Add test on cloud as circle
2) Add test on rects should be dense

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,164 @@
+            
+            bool IsNotIntersectedWithAnyAndExcludeSelf(Rectangle rect)
+            {
+               return circularCloudLayouter.Rectangles.Any(p => !p.IntersectsWith(rect) && p != rect);
```
Проверяем непересечение хотя бы с одним прямоугольником кроме себя, а по идее надо со всеми

Source34 replied:
Тут была ошибка в тесте (не со всеми сравнивалось), и в реализации размещения прямоугольников. Границы прямоугольников находились по оси находились на одной координате, и они пересекались. Сейчас прямоугольники размещаются с минимальным расстоянием между друг другом. Отрефакторен тест и сам метод размещения.

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,164 @@
+
+                area += rect.Height * rect.Width;
+
+                var currentRadius = GetDistance(circularCloudLayouter.Center, rect.GetCenter());
```
Кажется более правильным брать не центр прямоугольника, а самый отдалённый от центра угол

Danko19 replied:
Как вариант, можно потом, когда все прямоугольники уже нарисованы, взять самую левую, правую, верхнюю, нижнюю точку облака (через Min/Max) и сравнить их расстояние с центром - самое большое расстояние и будет радиусом

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,164 @@
+                    radius = currentRadius;
+            }
+
+            var expectedRadius = Math.Sqrt(area / Math.PI) * 1.2;
```
Немного не понял эту формулу, нужна пояснительная бригада 😅

Source34 replied:
Радиус через площадь считаем, с хитрым коэффициентом :)

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,164 @@
+            }
+
+            var expectedRadius = Math.Sqrt(area / Math.PI) * 1.2;
+            radius.Should().BeLessThan(expectedRadius);
```
Кажется не совсем корректным проверять форму облака через площади. У круга и, скажем, треугольника тоже могут быть разные площади, но форма у них разная

Появилась такая идея. Чтобы проверить форму облака, что она похожа на круг, можно взять эти же угловые точки (левую, правую, верхнюю, нижнюю) и сравнить их расстояние от центра круга между собой. По идее они должны быть примерно одинаковые с небольшим отклонением

Source34 replied:
++

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,164 @@
+                toCenterByX.Offset(new Point(Math.Sign(direction.X), 0));
+                toCenterByY.Offset(new Point(0, Math.Sign(direction.Y)));
+
+                toCenterByX.IntersectsWith(circularCloudLayouter.Rectangles)
```
А вот плотность лучше как раз проверять через площадь. Что сумма площадей прямоугольников занимает какой-то процент от площади круга (чем больше тем лучше)

Source34 replied:
++

Source34 committed:
Test fix and PutNextRectangle fix

1) CircularCloudLayouter method PutNextRectangle fix - the rectangles now do not intersect
2) Refactor CircularCloudLayouterTests

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,166 @@
+            var right = Math.Abs(circularCloudLayouter.Rectangles.Max(p => p.X) - 600);
+            var bottom = Math.Abs(circularCloudLayouter.Rectangles.Min(p => p.Y) - 600);
+            
+            var avgRadius = (int)((top + bottom + left + right) / (double)4);
```
Кажется более правильным взять не средний радиус, а максимальный, как считаешь?

Source34 replied:
Cогласен, исправил

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,166 @@
+            }
+
+            var top = circularCloudLayouter.Rectangles.Max(p => p.Top);
+            var left = circularCloudLayouter.Rectangles.Max(p => p.Left);
```
Для каких-то сторон нужно брать `Min`

Source34 replied:
Исправлено

Danko19 commented on cs/TagsCloudVisualization/TagsCloudVisualization.Core.Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,166 @@
+            for (var i = 0; i < rectsCount; i++)
+                circularCloudLayouter.PutNextRectangle(new Size(rnd.Next(10, 20), rnd.Next(15, 30)));
+
+            var tmpRects = new List<Rectangle>(circularCloudLayouter.Rectangles);
```
Почему не хочешь просто проверить, что каждый прямоугольник не пересекается ни с одним другим кроме себя? Да, будет n^2 проверок, но для теста оптимальность это не самое главное, главное чтобы проверки были исчерпывающие

Source34 replied:
Оно же по сути так и есть, только вместо проверки if(rect != self && ...) мы удаляем из коллекции текущий rect, и ищем пересечения со всеми остальными. Т.е. мы исключаем проверку на самого себя. В плане покрытия тут вроде норм, по итогу, проверяется каждый rect с каждым. 

Мб производительность будет лучше с if чем с remove() из листа, но это уже немного другая история).

Danko19 replied:
Код реализации "в лоб" с проверкой всех со всеми был бы проще и читался бы лучше, что для немаловажно для тестов. Но это не критично и скорее вкусовщина, докапываться не буду

Source34 committed:
Tests refactoring

1) Fix PutNextRectangle_WhenCorrectArgs_ShouldCloudAsCircle
2) Fix PutNextRectangle_WhenCorrectArgs_RectsShouldBeDense
